src/twisted/internet/udp.py

```diff
@@ -504,9 +516,33 @@ def _joinAddr2(self, interface, addr, join):
             return failure.Failure(error.MulticastJoinError(addr, interface, *e.args))
 
 
+    def _joinSourceGroup(self, source, interface, addr, join):
+        if not hasattr(socket, 'IP_ADD_SOURCE_MEMBERSHIP'):
+            setattr(socket, 'IP_ADD_SOURCE_MEMBERSHIP', 39)
+        if not hasattr(socket, 'IP_DROP_SOURCE_MEMBERSHIP'):
+            setattr(socket, 'IP_DROP_SOURCE_MEMBERSHIP', 40)
+
+        addr = socket.inet_aton(addr)
+        interface = socket.inet_aton(interface)
+        source = socket.inet_aton(source)
+        if join:
+            cmd = socket.IP_ADD_SOURCE_MEMBERSHIP
+        else:
+            cmd = socket.IP_DROP_SOURCE_MEMBERSHIP
+        try:
+            self.socket.setsockopt(socket.IPPROTO_IP, cmd, addr + interface + source)
```

Comment:
This socket option doesn't look right.  *addr + interface + source* is not the correct order for *struct mreq* which is what this parameter takes.  There are other examples out there which show how to use struct.pack for correctly packing the structure to be used in this socket option.  Here is one:

https://stackoverflow.com/questions/603852/multicast-in-python